[Bug]: Reports duplicate key value violates unique constraint error

### Is this a critical security issue?

- [x] This is not a security issue.

### Describe the Bug

It is possible for two updates to hit the same PuppetDB at the same time for the same server asset and create a postgresql unique constraint error/issue.

### Expected Behavior

If two transactions for the same resource hit postgresql at the same time, only one should be accepted/processed.

### Steps to Reproduce

1. Have two clusters of OpenVox Server behind a load balancer. (1 Global VIP with 2 local VIPs)
2. Have two clusters of OpenVoxDB behind a second load balancer. (1 Global VIP with 2 local VIPs)
3. Any update sent to an OpenVox Server should be sent to both OpenVoxDB instances at the same time.
4. But due to how the load balancer team configured the global load balancer, both updates go to the same OpenVoxDB server at the same time for the same resource.
5. Each OpenVox Agent run after that results in the same postgresql unique constrain error.

### Environment

Version: openvoxdb 8.9.1
Platform: AlmaLinux 9

Version: postgresql 14
Platform: Ubuntu 22.04

### Additional Context

Not sure how common this is out in the broader community, but I found it in two test environments.

### Relevant log output

```shell
2025-06-12T18:34:15.826Z INFO  [p.p.command] [23011-1749753255707-1749753255637] [109 ms] 'replace facts' command (d079c147) processed for myserver01.myorganization.org
2025-06-12T18:35:02.614Z ERROR [p.p.command] [23,012] [store report] Scheduling retry after attempt 0 for myserver01.myorganization.org
java.sql.BatchUpdateException: Batch entry 0 UPDATE reports SET latest = ('FALSE') WHERE latest = true and certname = ('myserver01.myorganization.org') was aborted: ERROR: duplicate key value violates unique constraint "reports_20250611z_encode_producer_timestamp_idx"
  Detail: Key (encode(hash, 'hex'::text), producer_timestamp)=(872f0e6ea94923b3372dcdcc8e85397b12374905, 2025-06-11 01:41:48.341+00) already exists.  Call getNextException to see other errors in the batch.
       	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:165)
       	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2413)
       	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:579)
       	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:912)
       	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:936)
       	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1733)
       	at com.zaxxer.hikari.pool.ProxyStatement.executeBatch(ProxyStatement.java:127)
       	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeBatch(HikariProxyPreparedStatement.java)
       	at clojure.java.jdbc$execute_batch.invokeStatic(jdbc.clj:598)
       	at clojure.java.jdbc$execute_batch.invoke(jdbc.clj:591)
       	at clojure.java.jdbc$db_do_execute_prepared_statement$fn__3649.invoke(jdbc.clj:1057)
       	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:860)
       	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:776)
       	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:789)
       	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:776)
       	at clojure.java.jdbc$db_do_execute_prepared_statement.invokeStatic(jdbc.clj:1056)
       	at clojure.java.jdbc$db_do_execute_prepared_statement.invoke(jdbc.clj:1042)
       	at clojure.java.jdbc$db_do_prepared.invokeStatic(jdbc.clj:1080)
       	at clojure.java.jdbc$db_do_prepared.invoke(jdbc.clj:1060)
       	at clojure.java.jdbc$execute_BANG_$execute_helper__3717.invoke(jdbc.clj:1464)
       	at clojure.java.jdbc$execute_BANG_.invokeStatic(jdbc.clj:1466)
       	at clojure.java.jdbc$execute_BANG_.invoke(jdbc.clj:1435)
       	at clojure.java.jdbc$update_BANG_.invokeStatic(jdbc.clj:1686)
       	at clojure.java.jdbc$update_BANG_.invoke(jdbc.clj:1673)
       	at puppetlabs.puppetdb.jdbc$update_BANG_.invokeStatic(jdbc.clj:150)
       	at puppetlabs.puppetdb.jdbc$update_BANG_.invoke(jdbc.clj:147)
       	at puppetlabs.puppetdb.scf.storage$fn__32542$add_report_BANG__STAR___32547$fn__32548$fn__32549$fn__32551$fn__32554.invoke(storage.clj:1497)
       	at puppetlabs.puppetdb.scf.storage$fn__32542$add_report_BANG__STAR___32547$fn__32548$fn__32549$fn__32551.invoke(storage.clj:1495)
       	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:860)
       	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:776)
       	at puppetlabs.puppetdb.scf.storage$fn__32542$add_report_BANG__STAR___32547$fn__32548$fn__32549.invoke(storage.clj:1454)
       	at puppetlabs.puppetdb.scf.storage.proxy$java.lang.Object$Callable$7da976d4.call(Unknown Source)
       	at com.codahale.metrics.Timer.time(Timer.java:101)
       	at puppetlabs.puppetdb.scf.storage$fn__32542$add_report_BANG__STAR___32547$fn__32548.invoke(storage.clj:1445)
       	at puppetlabs.puppetdb.scf.storage$fn__32542$add_report_BANG__STAR___32547.invoke(storage.clj:1436)
       	at puppetlabs.puppetdb.scf.storage$fn__32834$add_report_BANG___32843$fn__32849$store_BANG___32851$fn__32852.invoke(storage.clj:2000)
       	at puppetlabs.puppetdb.jdbc$retry_with_monitored_connection$fn__25387$fn__25400.invoke(jdbc.clj:479)
       	at clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:807)
       	at clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:776)
       	at puppetlabs.puppetdb.jdbc$retry_with_monitored_connection$fn__25387.invoke(jdbc.clj:475)
       	at puppetlabs.puppetdb.jdbc$retry_sql$attempt__25367.invoke(jdbc.clj:423)
       	at puppetlabs.puppetdb.jdbc$retry_sql.invokeStatic(jdbc.clj:433)
       	at puppetlabs.puppetdb.jdbc$retry_sql.invoke(jdbc.clj:414)
       	at puppetlabs.puppetdb.jdbc$retry_with_monitored_connection.invokeStatic(jdbc.clj:472)
       	at puppetlabs.puppetdb.jdbc$retry_with_monitored_connection.invoke(jdbc.clj:467)
       	at puppetlabs.puppetdb.scf.storage$fn__32834$add_report_BANG___32843$fn__32849$store_BANG___32851.invoke(storage.clj:1995)
       	at puppetlabs.puppetdb.scf.storage$fn__32834$add_report_BANG___32843$fn__32849.invoke(storage.clj:2002)
       	at puppetlabs.puppetdb.scf.storage$fn__32834$add_report_BANG___32843.invoke(storage.clj:1981)
       	at puppetlabs.puppetdb.command$exec_store_report.invokeStatic(command.clj:486)
       	at puppetlabs.puppetdb.command$exec_store_report.invoke(command.clj:481)
       	at clojure.lang.AFn.applyToHelper(AFn.java:171)
       	at clojure.lang.AFn.applyTo(AFn.java:144)
       	at clojure.core$apply.invokeStatic(core.clj:675)
       	at clojure.core$partial$fn__5908.doInvoke(core.clj:2639)
       	at clojure.lang.RestFn.invoke(RestFn.java:467)
       	at puppetlabs.puppetdb.command$exec_command.invokeStatic(command.clj:568)
       	at puppetlabs.puppetdb.command$exec_command.invoke(command.clj:552)
       	at puppetlabs.puppetdb.command$attempt_exec_command$fn__36399.invoke(command.clj:661)
       	at puppetlabs.puppetdb.command$call_with_quick_retry$fn__36388.invoke(command.clj:641)
       	at puppetlabs.puppetdb.command$call_with_quick_retry.invokeStatic(command.clj:639)
       	at puppetlabs.puppetdb.command$call_with_quick_retry.invoke(command.clj:637)
       	at puppetlabs.puppetdb.command$attempt_exec_command.invokeStatic(command.clj:659)
       	at puppetlabs.puppetdb.command$attempt_exec_command.invoke(command.clj:655)
       	at puppetlabs.puppetdb.command$process_cmd$fn__36455$fn__36456.invoke(command.clj:787)
       	at puppetlabs.puppetdb.command$process_cmd$fn__36455.invoke(command.clj:784)
       	at puppetlabs.puppetdb.utils.metrics$multitime_BANG__STAR_$fn__31132$fn__31133$fn__31134.invoke(metrics.clj:23)
       	at puppetlabs.puppetdb.utils.metrics.proxy$java.lang.Object$Callable$7da976d4.call(Unknown Source)
       	at com.codahale.metrics.Timer.time(Timer.java:101)
       	at puppetlabs.puppetdb.utils.metrics$multitime_BANG__STAR_$fn__31132$fn__31133.invoke(metrics.clj:23)
       	at puppetlabs.puppetdb.utils.metrics$multitime_BANG__STAR_$fn__31132$fn__31133$fn__31134.invoke(metrics.clj:23)
       	at puppetlabs.puppetdb.utils.metrics.proxy$java.lang.Object$Callable$7da976d4.call(Unknown Source)
       	at com.codahale.metrics.Timer.time(Timer.java:101)
       	at puppetlabs.puppetdb.utils.metrics$multitime_BANG__STAR_$fn__31132$fn__31133.invoke(metrics.clj:23)
       	at puppetlabs.puppetdb.utils.metrics$multitime_BANG__STAR_.invokeStatic(metrics.clj:26)
       	at puppetlabs.puppetdb.utils.metrics$multitime_BANG__STAR_.invoke(metrics.clj:15)
       	at puppetlabs.puppetdb.command$process_cmd.invokeStatic(command.clj:781)
       	at puppetlabs.puppetdb.command$process_cmd.invoke(command.clj:767)
       	at puppetlabs.puppetdb.command$process_message.invokeStatic(command.clj:880)
       	at puppetlabs.puppetdb.command$process_message.invoke(command.clj:830)
       	at puppetlabs.puppetdb.command$message_handler$fn__36510.invoke(command.clj:925)
       	at puppetlabs.puppetdb.threadpool$dochan$fn__36066$fn__36067.invoke(threadpool.clj:115)
       	at puppetlabs.puppetdb.threadpool$gated_execute$fn__36027.invoke(threadpool.clj:68)
       	at clojure.lang.AFn.run(AFn.java:22)
       	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
       	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
       	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "reports_20250611z_encode_producer_timestamp_idx"
  Detail: Key (encode(hash, 'hex'::text), producer_timestamp)=(872f0e6ea94923b3372dcdcc8e85397b12374905, 2025-06-11 01:41:48.341+00) already exists.
       	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2725)
       	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2412)
       	... 84 common frames omitted
```

--- a/src/puppetlabs/puppetdb/scf/storage.clj
+++ b/src/puppetlabs/puppetdb/scf/storage.clj
@@ -1494,7 +1494,7 @@
                      ;; from the case where the incoming report needs partition(s) made for it.
                      (try
                        (jdbc/update! "resource_events" {:latest false} ["latest = true and report_id = ?" latest-report-id])
-                       (jdbc/update! "reports" {:latest false} ["latest = true and certname = ?" certname])
+                       (jdbc/update! "reports" {:latest false} ["latest = true and id = ?" latest-report-id])
                        (catch java.sql.BatchUpdateException e
                          ;; undefined table -- inheritance partitions, or declarative partitions w/ no children
                          ;; check constraint violation -- declarative partitions when row does not match child partitions

